rita/src/rita_common/tunnel_manager/mod.rs

```diff
@@ -38,33 +44,75 @@ type Resolver = Mocker<resolver::Resolver>;
 #[cfg(not(test))]
 type Resolver = resolver::Resolver;
 
-#[derive(Debug, Fail)]
-pub enum TunnelManagerError {
-    #[fail(display = "DNS lookup error")]
-    DNSLookupError,
+/* Uncomment when tunnel state handling is added
+#[derive(Debug, Clone)]
+pub enum TunnelState {
+    Init,
+    Open,
+    Throttled,
+    Closed,
 }
+*/
 
 #[derive(Debug, Clone)]
-pub struct TunnelData {
+pub struct Tunnel {
+    pub ip: IpAddr,
     pub iface_name: String,
+    pub listen_ifidx: u32,
     pub listen_port: u16,
+    //    pub tunnel_state: TunnelState,
+    pub localid: LocalIdentity,
 }
 
-impl TunnelData {
-    fn new(listen_port: u16) -> TunnelData {
+impl Tunnel {
+    fn new(
+        ip: IpAddr,
+        our_listen_port: u16,
+        ifidx: u32,
+        their_id: LocalIdentity,
+    ) -> Result<Tunnel, Error> {
         let iface_name = KI.setup_wg_if().unwrap();
-        TunnelData {
-            iface_name,
-            listen_port,
-        }
+
+        //let init = TunnelState::Init;
+        let tunnel = Tunnel {
+            ip: ip,                       //tunnel endpoint
+            iface_name: iface_name,       //name of wg#
+            listen_ifidx: ifidx,          //The physical port this tunnel is listening on
+            listen_port: our_listen_port, //the port this tunnel resides on
+            //            tunnel_state: init,           //how this tunnel feels about it's life
+            localid: their_id.clone(), // the identity of the counterparty tunnel once we have it
+        };
+
+        let network = SETTING.get_network().clone();
+
+        // TODO you have the iface index use it
+        KI.open_tunnel(
+            &tunnel.iface_name,
+            tunnel.listen_port,
+            &SocketAddr::new(ip, their_id.wg_port),
+            &their_id.global.wg_public_key,
+            Path::new(&network.wg_private_key_path),
+            &network.own_ip,
+            network.external_nic.clone(),
+            &mut SETTING.get_network_mut().default_route,
+        )?;
+
+        let stream = TcpStream::connect::<SocketAddr>(
+            format!("[::1]:{}", SETTING.get_network().babel_port).parse()?,
+        )?;
+
+        let mut babel = Babel::new(stream);
+
+        babel.start_connection()?;
+        babel.monitor(&tunnel.iface_name)?;
+
+        Ok(tunnel)
     }
 }
 
 pub struct TunnelManager {
-    pub port: u16,
-
-    tunnel_map: HashMap<IpAddr, TunnelData>,
-    listen_interfaces: HashSet<String>,
+    free_ports: Vec<u16>,
+    tunnels: Vec<Tunnel>,
```

Comment:
We most likely want sets for these

Reply:
don't see the advantage for port, we never search and never check membership. No duplicates maybe,

Reply:
nah hashset doesn't make sense for ports, no push or pop makes getting a port (we don't care which one) difficult.